testcase/unittest/blockchain/transactions/test_transaction_serializer.py

```diff
@@ -0,0 +1,280 @@
+import pytest
+from freezegun import freeze_time
+
+from loopchain.blockchain.transactions import TransactionSerializer, TransactionVersioner, Transaction
+from loopchain.blockchain.transactions import genesis, v2, v3, v3_issue
+from loopchain.blockchain.types import Hash32
+
+tx_versioner = TransactionVersioner()
+
+
+class TestTransactionSerializerBase:
+    @pytest.mark.parametrize("tx_version", [genesis.version, v2.version, v3.version])
+    def test_serializer_version_check(self, tx_factory, tx_version):
+        tx: Transaction = tx_factory(tx_version=tx_version)
+        ts = TransactionSerializer.new(version=tx.version, type_=tx.type(), versioner=tx_versioner)
+
+        if tx_version == genesis.version:
+            assert isinstance(ts, genesis.TransactionSerializer)
+        elif tx_version == v2.version:
+            assert isinstance(ts, v2.TransactionSerializer)
+        elif tx_version == v3.version:
+            assert isinstance(ts, v3.TransactionSerializer)
+        else:
+            assert False
+
+    @pytest.mark.parametrize("tx_version, expected_keys", [
+        (genesis.version, ["accounts", "message", "nid", "signature"]),
+        (v2.version, ["from", "to", "value", "fee", "timestamp", "nonce"]),
+        (v3.version, ["version", "from", "to", "stepLimit", "timestamp", "nid", "value", "nonce"])
+    ])
+    def test_to_origin_data_has_valid_form(self, tx_factory, tx_version, expected_keys):
+        tx: Transaction = tx_factory(tx_version=tx_version)
+        ts = TransactionSerializer.new(version=tx.version, type_=tx.type(), versioner=tx_versioner)
+
+        origin_data = ts.to_origin_data(tx)
+        assert set(origin_data) == set(expected_keys)
+
+
+class TestTransactionSerializer_genesis:
+    tx_version = genesis.version
+
+    # TODO: What is diffrence between tx.raw_data and its dict casted?
```

Comment:
What is diffrence between tx.raw_data and its dict casted?

Reply:
It is dangerous to return `raw_data` without copying.
Assume that `raw_data` is manipulated out of transaction. Then the transaction will lose integrity.

Reply:
I see.

Reply:
-> fixed by casting raw_data to dict:
https://github.com/icon-project/loopchain/blob/92b74a99952fa92f12caa81fad8f648f4ca55524/loopchain/blockchain/transactions/genesis/transaction_serializer.py#L19